File: Problem 2 design hashset.java
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class MyHashSet {

    List<Integer> hashSet;
    
    /** Initialize your data structure here. */
    public MyHashSet() {
        hashSet = new ArrayList<Integer>();
    }
    
    public void add(int key) {
        if(hashSet.contains(key)) {
            return;
        } else {
            hashSet.add(key);
        }
    }
    
    public void remove(int key) {
        if(!hashSet.contains(key)) return;
        hashSet.remove(hashSet.indexOf(key));
    }
    
    /** Returns true if this set contains the specified element */
    public boolean contains(int key) {
        return hashSet.contains(key);
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */

File: Problem2_DesignHashset.py
# Time Complexity: O(1)
# Space Complexity: O(n)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

# Solution:

class MyHashSet:

    # Initialize primary and secondary array size (square root of total no of keys)
    def __init__(self):
        self.buckets = 1000
        self.bucket_items = 1001
        self.storage = [None] * self.buckets

    # Returns the index of the bucket
    def bucket(self, key) -> int:
        return int(key % self.buckets)

    # Returns the index of the bucket_items
    def bucket_item(self, key) -> int:
        return int(key / self.bucket_items)

    # add - Set the value of the computed hash address to True
    def add(self, key: int) -> None:
        bucket = self.bucket(key)
        bucket_item = self.bucket_item(key)

        if self.storage[bucket] == None:
            self.storage[bucket] = [False] * self.bucket_items

        self.storage[bucket][bucket_item] = True

    # remove - Set the value of the computed hash address to False
    def remove(self, key: int) -> None:
        bucket = self.bucket(key)
        bucket_item = self.bucket_item(key)

        if self.storage[bucket] == None:
            return

        self.storage[bucket][bucket_item] = False

    # contains - Return the value of the computed hash address - True or False
    def contains(self, key: int) -> bool:
        bucket = self.bucket(key)
        bucket_item = self.bucket_item(key)

        if self.storage[bucket] == None:
            return False

        return self.storage[bucket][bucket_item]
